llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp

```diff
@@ -0,0 +1,148 @@
+                : std::make_unique<MachineIRBuilder>();
+    Builder->setMF(MF);
+    MachineIRBuilder &MIB = *Builder;
+    // Set Observer to inform change of register type or change in op-code
+    GISelObserverWrapper Observer;
+    if (CSEInfo) {
+      Observer.addObserver(CSEInfo);
+      MIB.setChangeObserver(Observer);
+    }
+
+    bool Changed = false;
+    for (MachineBasicBlock &MBB : MF) {
+      // If a BB has PHI nodes, the very first instruction will be PHI
+      if (MBB.empty() ||
+          MBB.begin()->getOpcode() != llvm::TargetOpcode::G_PHI) {
+        continue;
+      }
+      Changed = processMBBWithPHI(MBB, MRI, MIB, Observer);
+    }
+    return Changed;
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<MachineModuleInfoWrapperPass>();
+    AU.addRequired<GISelCSEAnalysisWrapperPass>();
+    AU.addRequired<TargetPassConfig>();
+    AU.setPreservesAll();
+  }
+
+  StringRef getPassName() const override { return AIE_ELIMINATE_DUPLICATE_PHI; }
+
+private:
+  void eliminateDuplicatePHIUses(MachineRegisterInfo &MRI,
+                                 GISelObserverWrapper &Observer,
+                                 Register BaseReg, Register DuplicateReg) {
+    const auto UseInstIter =
+        make_early_inc_range(MRI.use_nodbg_instructions(DuplicateReg));
+    for (auto &UseInstr : UseInstIter) {
+      for (auto &MO : UseInstr.uses()) {
+        if (MO.isReg() && MO.getReg() == DuplicateReg) {
+          Observer.changingInstr(UseInstr);
+          MO.setReg(BaseReg);
+          Observer.changedInstr(UseInstr);
+        }
+      }
+    }
+  }
+
+  bool isPointerOperandPHI(MachineRegisterInfo &MRI, const MachineInstr &MI) {
+    assert(MI.getOpcode() == TargetOpcode::G_PHI);
+    const Register Reg = MI.getOperand(0).getReg();
+    const LLT RegType = MRI.getType(Reg);
+    return RegType == LLT::pointer(DefaultAddrSpace, RegType.getSizeInBits());
+  }
+
+  bool processMBBWithPHI(MachineBasicBlock &MBB, MachineRegisterInfo &MRI,
+                         MachineIRBuilder &MIB,
+                         GISelObserverWrapper &Observer) {
+    bool Changed = false;
+    auto LastPHI = MBB.getFirstNonPHI();
+    for (auto &MI : make_early_inc_range(make_range(MBB.begin(), LastPHI))) {
+      if (!isPointerOperandPHI(MRI, MI))
+        continue;
+      for (auto &PHI : make_early_inc_range(
+               make_range(std::next(MachineBasicBlock::instr_iterator(MI)),
+                          MachineBasicBlock::instr_iterator(*LastPHI)))) {
+        if (!isPointerOperandPHI(MRI, PHI))
+          continue;
+        if (MI.isIdenticalTo(PHI, MachineInstr::IgnoreDefs)) {
+          LLVM_DEBUG(dbgs() << "Identical PHI nodes found\n");
+          LLVM_DEBUG(dbgs() << "  Base PHI: " << MI);
+          LLVM_DEBUG(dbgs() << "  Duplicate PHI: " << PHI);
+          eliminateDuplicatePHIUses(MRI, Observer, MI.getOperand(0).getReg(),
+                                    PHI.getOperand(0).getReg());
```

Comment:
Checking: I guess you can't really eliminate `PHI`, because it might be the next instruction that the outer loop iterates on?

Reply:
Can we provide an example I am not very sure ?

Reply:
Reading the name, I was expecting that the function deletes `PHI` from the basic block. But it "only" retargets its uses to `MI`, which I think is safer.